File: solutions/python/year2022/day02.py
from enum import Enum
from typing import TypeVar, overload


class Shape(Enum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3


class RoundResult(Enum):
    LOSS = 0
    DRAW = 3
    WIN = 6


class OponentShape(Enum):
    ROCK = "A"
    PAPER = "B"
    SCISSORS = "C"


# Part 1: X, Y and Z are the hand shapes that we should play each round.
class PlayerShape(Enum):
    ROCK = "X"
    PAPER = "Y"
    SCISSORS = "Z"

    @property
    def points(self) -> int:
        return Shape[self.name].value

    def result_against(self, oponent_shape: OponentShape) -> RoundResult:
        match (self, oponent_shape):
            case (
                (self.ROCK, OponentShape.SCISSORS)
                | (self.PAPER, OponentShape.ROCK)
                | (self.SCISSORS, OponentShape.PAPER)
            ):
                return RoundResult.WIN

            case (
                (self.SCISSORS, OponentShape.ROCK)
                | (self.PAPER, OponentShape.SCISSORS)
                | (self.ROCK, OponentShape.PAPER)
            ):
                return RoundResult.LOSS

            case _:
                return RoundResult.DRAW


# Part 2: X, Y and Z are the results that we should enforce each round.
class PlayerResult(Enum):
    LOSS = "X"
    DRAW = "Y"
    WIN = "Z"

    @property
    def points(self) -> int:
        return RoundResult[self.name].value

    def necessary_shape_against(self, oponent_shape: OponentShape) -> Shape:
        match (self, oponent_shape):
            case (
                (self.LOSS, oponent_shape.PAPER)
                | (self.DRAW, oponent_shape.ROCK)
                | (self.WIN, oponent_shape.SCISSORS)
            ):
                return Shape.ROCK

            case (
                (self.LOSS, oponent_shape.SCISSORS)
                | (self.DRAW, oponent_shape.PAPER)
                | (self.WIN, oponent_shape.ROCK)
            ):
                return Shape.PAPER

            case _:
                return Shape.SCISSORS


P = TypeVar("P", PlayerShape, PlayerResult)
Game = list[tuple[OponentShape, P]]


@overload
def parse_game(input: str, parser: type[PlayerShape]) -> Game[PlayerShape]:
    ...


@overload
def parse_game(input: str, parser: type[PlayerResult]) -> Game[PlayerResult]:
    ...


def parse_game(
    input: str, parser: type[PlayerShape] | type[PlayerResult]
) -> Game[PlayerResult] | Game[PlayerShape]:
    game = []
    for round in input.strip().split("\n"):
        oponent, player = round.split(" ")
        game.append((OponentShape(oponent), parser(player)))

    return game


def part1(game: Game[PlayerShape]) -> int:
    score = 0
    for oponent, player in game:
        score += player.points + player.result_against(oponent).value

    return score


def part2(game: Game[PlayerResult]) -> int:
    score = 0
    for oponent, result in game:
        score += result.points + result.necessary_shape_against(oponent).value

    return score


def solution(input: str) -> tuple[int, int]:
    return (
        part1(parse_game(input, PlayerShape)),
        part2(parse_game(input, PlayerResult)),
    )

File: solutions/c/src/year2022/day02.c
typedef enum {
    ROCK = 1,
    PAPER = 2,
    SCISSORS = 3,
} Shape;

typedef enum {
    LOSS = 0,
    DRAW = 3,
    WIN = 6,
} RoundResult;

Shape parse_oponent(char encoded) {
    switch (encoded) {
        case 'A': return ROCK;
        case 'B': return PAPER;
        case 'C': return SCISSORS;
        default:  return -1;
    }
}

Shape parse_player_shape(char encoded) {
    switch (encoded) {
        case 'X': return ROCK;
        case 'Y': return PAPER;
        case 'Z': return SCISSORS;
        default:  return -1;
    }
}

RoundResult parse_player_result(char encoded) {
    switch (encoded) {
        case 'X': return LOSS;
        case 'Y': return DRAW;
        case 'Z': return WIN;
        default:  return -1;
    }
}

RoundResult round_result(Shape player, Shape oponent) {
    switch (player) {
        case ROCK: switch (oponent) {
            case ROCK: return DRAW;
            case PAPER: return LOSS;
            case SCISSORS: return WIN;
        }

        case PAPER: switch (oponent) {
            case ROCK: return WIN;
            case PAPER: return DRAW;
            case SCISSORS: return LOSS;
        }

        case SCISSORS: switch (oponent) {
            case ROCK: return LOSS;
            case PAPER: return WIN;
            case SCISSORS: return DRAW;
        }
    }
}

Shape necessary_shape_for_result(Shape oponent, RoundResult result) {
    switch (result) {
        case LOSS: switch (oponent) {
            case ROCK: return SCISSORS;
            case PAPER: return ROCK;
            case SCISSORS: return PAPER;
        }

        case WIN: switch (oponent) {
            case ROCK: return PAPER;
            case PAPER: return SCISSORS;
            case SCISSORS: return ROCK;
        }

        case DRAW: return oponent;
    }
}

void year2022_day02_solution(char *input, char **output) {
    int part1_total_points = 0;
    int part2_total_points = 0;

    while (*input != '\0') {
        Shape oponent_shape = parse_oponent(*input);
        input += 2;

        Shape player_shape = parse_player_shape(*input);
        RoundResult player_result = parse_player_result(*input);
        input += 2;

        part1_total_points += round_result(player_shape, oponent_shape) + player_shape;
        part2_total_points += necessary_shape_for_result(oponent_shape, player_result) + player_result;
    }

    sprintf(output[0], "%d", part1_total_points);
    sprintf(output[1], "%d", part2_total_points);
}
